### ISSUE TITLE
[connectivity_plus][iOS] checkConnectivity returns None unexpectedly occasionally

### ISSUE THREAD

**Issue by amantoux** [fluttercommunity/plus_plugins#858]
Since 2.3.0, connectivity_plus sometimes returns `ConnectivityResult.none` inappropriately.

I suspect it is related to the introduction of `PathMonitorConnectivityProvider` for iOS version >= 12

Comparing with `ReachabilityConnectivityProvider`, that latter uses by default the `.default` QoS queue whereas `PathMonitorConnectivityProvider` uses `.background` QoS queue.

Maybe the use of `.default` would be better

https://github.com/fluttercommunity/plus_plugins/blob/89993a6339cf74594e88bc48f861d0c95e8b21ea/packages/connectivity_plus/connectivity_plus/ios/Classes/PathMonitorConnectivityProvider.swift#L7

**Comment by AliRezaBeitari** [comment]
Might be related: https://github.com/fluttercommunity/plus_plugins/issues/852

**Comment by jfahrenkrug** [comment]
I'm also seeing weird behavior since 2.3.0. It's almost like it's flipped: When I turn wifi ON it returns "none", when I turn wifi OFF, it returns "wifi".

**Comment by comESK** [comment]
I'm seeing it now on an iOS 15.5 physical device, not just in the Simulator, so idk.

**Comment by alperefesahin** [comment]
the same issue is here in the last days of 2022.

**Comment by anchals333** [comment]
I'm facing the same issue in android & iOS physical devices. This issue usually comes when we move the app from background to foreground then even the internet is connected the connectivity status returns none.

**Comment by poojadev** [comment]
I'm still facing the same issue on Ios "checkConnectivity returns None unexpectedly occasionally"  anyone have any solution for it?

**Comment by comESK** [comment]
If you need a solution in the meantime, check out [internet_connection_checker](https://github.com/RounakTadvi/internet_connection_checker). They even have this code block in their README, for easy migration:
```dart
var isDeviceConnected = false;

var subscription = Connectivity().onConnectivityChanged.listen((ConnectivityResult result) async {
  if(result != ConnectivityResult.none) {
    isDeviceConnected = await InternetConnectionChecker().hasConnection;
  }
});
```

Personally, I still had issue on the first check on iOS, so I added in a delayed second check before showing the user an error. It's not the cleanest code I've written, but it's better than not having the feature at all.

Hope that helps while we wait!

**Comment by adminant** [comment]
Still, in 2023 the problem persists on iOS 15.1.
I switch off wifi on iPhone, the listener comes with ConnectivityResult.none, it's ok.
Then, I switch on wifi, and during a couple of seconds the ConnectivityResult.wifi arrives, then ConnectivityResult.none, then again ConnectivityResult.wifi. So all my app logic is ruined. What should I do?

**Comment by gmjgh** [comment]
To workaround the issue, you can add an empty `onConnectivityChanged` listener somewhere in your code - after this, the method `Connectivity().checkConnectivity()` will return a correct value.

**Comment by Saurabh-Detharia** [comment]
Any update on this? When I check the status first time on iOS, it's always giving none and then the next time, it works fine.

**Comment by praxder** [comment]
This is still an issue

**Comment by LeonardoRosaa** [comment]
> I'm seeing it now on an iOS 15.5 physical device, not just in the Simulator, so idk.

Same here. Some users have this on physical devices as well.

**Comment by prabakaranios** [comment]
`//    if #available(iOS 12, *) {
//      connectivityProvider = PathMonitorConnectivityProvider()
//    } else {
      connectivityProvider = ReachabilityConnectivityProvider()
//    }`

It will work when you use the `ReachabilityConnectivityProvider` provider.  because it's having a bug in the `PathMonitorConnectivityProvider` class.  

below are steps, you can take to reproduce the bug.
1. place the breakpoint.
2. Initially you will get `unsatisfied`. so you will get none
3. after that it's getting `satisfied` and you will get a valid status. 

Team, please fix the bug.

**Comment by miquelbeltran** [comment]
There is a PR that is doing some changes on iOS connectivity code. Anyone would like to help to review it? https://github.com/fluttercommunity/plus_plugins/pull/2599 I'd like to know if this issue could be resolved there too.

**Comment by iamsharad98** [comment]
I have used this as temporary fix, 
`
  Future<void> _updateConnectionStatus(ConnectivityResult result) async {
    if(_connectionStatus == ConnectivityResult.none
        && result == ConnectivityResult.none){
      return;
    }
    _connectionStatus = result;
    try {
      final result = await InternetAddress.lookup('google.com');
      if (result.isNotEmpty && result[0].rawAddress.isNotEmpty) {
        print('connected');
        _connectionStatus = ConnectivityResult.wifi;
      }
    } on SocketException catch (_) {
      print('not connected');
    }

    // print("_updateConnectionStatus: _connectionStatus $_connectionStatus");
    if (_connectionStatus == ConnectivityResult.none) {
      hasInterNetConnection = false;
        MessageDialogue().dialogueActionableOk("No internet connection",
            "Please make sure Wi-Fi or mobile data is turned on, then try again.",
                () async {
              await initConnectivity();
              if(hasInterNetConnection){
                Get.back();
              }
            }, actionText: "Retry");
    } else {
      hasInterNetConnection = true;
    }
    if (mounted) {
      setState(() {});
    }
  }`

**Comment by miquelbeltran** [comment]
I've been reading through the thread, and it's hard to follow what the actual issue is and how to reproduce it. Anyone facing this issue still, I'd recommend creating a new ticket providing up-to-date information on how to reproduce the issue.

---

Regarding the suggestion on changing the `DispatchQueue` priority: This is something that I am open to explore, at the moment the plugin uses `.background`.

Looking at the available ones: https://developer.apple.com/documentation/dispatch/dispatchqos what I have seen if some examples [link](https://gist.github.com/yosshi4486/0ecfcfcf44c14b25a7cf1b7c0fd72d49) using `.utility`, which seems a better option, and I agree that probably `.background` is too low.

I'll probably do a PR with that.

**Comment by miquelbeltran** [comment]
After doing some tests with QoS `.utility` and `.userInitiated` I cannot see any differences with `.background`.

What can be observed is that `NWPathMonitor` is just unreliable. I'd be getting `NWPath` saying it is "satisfied" even when having no network, and then getting "unsatisfied" when reconnecting, then immediately getting a new callback with a "satisfied" one. That's why some users report getting "none" and then "wifi" immediately, and I could observe this both with `.background` and `.userInitiated`.

This, to me, doesn't look like a problem with the plugin, but rather a problem of how `NWPathMonitor` is not very reliable.

The documentation already mentions:

> You should not be using the current network status for deciding whether you can reliably make a network connection.

I will add some comments about the unreliability of the iOS API to the README, there is not much more we can do IMO.

**Comment by pureimpro** [comment]
Same issue with my app, on Android with latest connectivity_plus version. Randomly happen when resuming app from background whereas wifi is still activated on Android side.
Stuck with connectivityStatus.offline.
Need to disable wifi and re-enable it to recover network from the plugin

**Comment by sohrabonline** [comment]
The same issue is still active.
